### PR TITLE
Extend torch.cuda.is_available() to attempt an NVML-based CUDA availability assessment when explicitly requested by the user

### LINKED ISSUE #83973: Implement `torch.cuda.device_count` without poison
### 🚀 The feature, motivation and pitch

Currently, when a user calls the functions `torch.cuda.device_count` or `torch.cuda.is_available`, PyTorch initializes some CUDA context that prevents us to fork processes. We get the well known error message "Cannot re-initialize CUDA in forked subprocess" as demonstrated in the code below:

```py
import torch 

def foo(idx):
    print(idx)
    torch.cuda.set_device(idx)

torch.cuda.device_count()  # this call causes an error: "Cannot re-initialize CUDA in forked subprocess"
assert not torch.cuda.is_initialized()
torch.multiprocessing.start_processes(foo, nprocs=2, start_method="fork")
```

In general, the error is accurate because it is not very meaningful to fork processes when CUDA memory is already allocated (e.g., tensors on device, model moved to device, etc.). However, here we are dealing with simple calls to utility functions `torch.cuda.device_count` that do not require any real initialization of the device or memory allocation. There does not seem to be any obvious reason why a call to this function should prevent us from forking processes. 

With this issue, we'd like to request that this behavior be changed.

Looking at the code, we find that the Python API calls into the C++:

https://github.com/pytorch/pytorch/blob/fa241fd50e226d74af06c36482fa68f0e2a4fb3c/torch/cuda/__init__.py#L459-L464

And its implementation is here

https://github.com/pytorch/pytorch/blob/fa241fd50e226d74af06c36482fa68f0e2a4fb3c/torch/csrc/cuda/Module.cpp#L120-L125

where we can find the [`poison_fork()`](https://github.com/pytorch/pytorch/blob/fa241fd50e226d74af06c36482fa68f0e2a4fb3c/torch/csrc/cuda/Module.cpp#L58) call that prevents forked processes to re-initialize CUDA.

**Options to consider:**
- Remove the poison_fork() for the device_count function, as there is no obvious reason why it has to be there.
- Provide a function to un-initialize the CUDA context and memory so that processes can be forked

### Alternatives

In [Lightning](https://github.com/Lightning-AI/lightning), we need to call `torch.cuda.device_count` early on in the Trainer without knowing whether processes are going to be forked or not later on. We currently avoid this issue by running the call to `torch.cuda.device_count` in a subprocess like this:

```
def num_cuda_devices():
    with multiprocessing.Pool(1) as pool:
        return pool.apply(torch.cuda.device_count)
```

This way, the CUDA context gets created in the subprocess while the main process does not get "poisoned". This workaround is not ideal because launching a new process just to evaluate this function is slow. 

### Additional context

https://github.com/Lightning-AI/lightning/issues/14292
https://github.com/Lightning-AI/lightning/pull/13405

cc @ngimel @carmocca @justusschock

### PR DESCRIPTION
**Pull request by speediedan**
Fixes #83973 (This is a substitute PR for https://github.com/pytorch/pytorch/pull/85024)

First of all, thanks for your invaluable contributions to PyTorch everyone!

Given how extensively `torch.cuda.is_available` is used in the PyTorch ecosystem, IMHO it's worthwhile to provide downstream libraries/frameworks/users the ability to alter the default behavior of `torch.cuda.is_available` in the context of their PyTorch usage. 

I'm confident there are many current and future such use cases which could benefit from leveraging a weakened, NVML-based `torch.cuda.is_available` assessment at a downstream framework's explicit direction (thanks @malfet https://github.com/pytorch/pytorch/commit/81da50a972fc402a6dd880fe392af0f0051cb6de !). Though one could always patch out the `torch.cuda.is_available` function with another implementation in a downstream library, I think this environmental variable based configuration option is more convenient and the cost to including the option is quite low.

As discussed in https://github.com/pytorch/pytorch/pull/85024#issuecomment-1261542045, this PR gates new non-default NVML-based CUDA behavior with an environmental variable (PYTORCH_NVML_BASED_CUDA_CHK) that allows a user/framework to invoke non-default, NVML-based `is_available()` assessments if desired.

Thanks again for your work everyone!
@ngimel @malfet @awaelchli

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -79,13 +79,24 @@ def _is_compiled() -> bool:
...
     r"""Returns a bool indicating if CUDA is currently available."""
     if not _is_compiled():
         return False
-    # This function never throws and returns 0 if driver is missing or can't
-    # be initialized
-    return torch._C._cuda_getDeviceCount() > 0
+    if _nvml_based_avail():
+        # The user has set an env variable to request this availability check that attempts to avoid fork poisoning by
+        # using NVML at the cost of a weaker CUDA availability assessment. Note that if NVML discovery/initialization
+        # fails, this assessment fails back to the default CUDA Runtime API assessment (`cudaGetDeviceCount`)
```
Comment (awaelchli) [inline torch/cuda/__init__.py:92]:
```suggestion
        # fails, this assessment falls back to the default CUDA Runtime API assessment (`cudaGetDeviceCount`)
```
Was it mean to say "fall back"? not sure

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -79,13 +79,24 @@ def _is_compiled() -> bool:
     r"""Returns true if compile with CUDA support."""
     return hasattr(torch._C, '_cuda_getDeviceCount')
 
+def _nvml_based_avail() -> bool:
+    return os.getenv('PYTORCH_NVML_BASED_CUDA_CHK') == '1'
```
Comment (awaelchli) [inline torch/cuda/__init__.py:83]: is the missing "e" in "check" a typo or intentional for abbreviation?

File: test/test_cuda_nvml_based_avail.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    # `PYTORCH_NVML_BASED_CUDA_CHK` is set to 1, reverting to the default CUDA Runtime API check otherwise.
+    # If the NVML-based assessment is attempted but fails, the CUDA Runtime API check should be executed
+    @unittest.skipIf(IS_WINDOWS, "Needs fork")
+    @parametrize("nvml_avail", [True, False])
+    @parametrize("avoid_init", ['1', '0', None])
+    def test_cuda_is_available(self, avoid_init, nvml_avail):
+        if avoid_init:
+            os.environ['PYTORCH_NVML_BASED_CUDA_CHK'] = avoid_init
+        else:
+            del os.environ['PYTORCH_NVML_BASED_CUDA_CHK']
```
Comment (awaelchli) [inline test/test_cuda_nvml_based_avail.py:46]: I think it would be better if you used monkeypatch or similar to ensure the env var does not stay past the test execution. I'm not sure how pytorch runs the test suite, but to be safe that this won't interact with other tests.

**Review by awaelchli — COMMENTED** [review]
Great!

File: torch/cuda/__init__.py (hunk 1 above)
Comment (speediedan) [reply torch/cuda/__init__.py:92]: good suggestion, falls back is a bit better I agree.

File: torch/cuda/__init__.py (hunk 2 above)
Comment (speediedan) [reply torch/cuda/__init__.py:83]: I had intentionally abbreviated CHECK to CHK because the var seemed to be getting a bit long already but I'm happy to change it to CHECK if you guys think that's better. Just lmk...

File: test/test_cuda_nvml_based_avail.py (hunk 3 above)
Comment (speediedan) [reply test/test_cuda_nvml_based_avail.py:46]: Though these tests will run in subprocesses during CI (configured via run_tests.py), 
https://github.com/pytorch/pytorch/blob/8669f6d42691c2124414cc97d0061ea6a0143007/test/run_test.py#L416
I agree it's better to ensure the env var scope is narrowed as much as possible. I'll push the context managed patch version shortly, thanks!

File: docs/source/notes/cuda.rst [hunk 4]
```diff
@@ -466,6 +466,22 @@ have a flag that can be used to disable CUDA, in combination with
...
+    When assessing the availability of CUDA in a given environment (:meth:`~torch.cuda.is_available`), PyTorch's default
+    behavior is to call the CUDA Runtime API method `cudaGetDeviceCount`_. Because this call in turn initializes the
+    CUDA Driver API (via `cuInit`_) if it is not already initialized, subsequent forks of a process that has run
+    :meth:`~torch.cuda.is_available` will fail with a CUDA initialization error.
+
+    One can set ``PYTORCH_NVML_BASED_CUDA_CHECK=1`` in your environment before importing PyTorch modules that execute
+    :meth:`~torch.cuda.is_available` (or before executing it directly) in order to direct
+    :meth:`~torch.cuda.is_available` to attempt an NVML-based assessement (`nvmlDeviceGetCount_v2`_). If the
+    NVML-based assessment is successful (i.e. NVML discovery/initialization does not fail),
+    :meth:`~torch.cuda.is_available` calls will not poison subsequent forks.
```
Comment (ngimel) [inline docs/source/notes/cuda.rst:480]: Can you add that this is a less reliable method of determining if cuda is available, because nvml-based check can succeed, and yet pytorch won't be able to initialize cuda context for other reasons.

File: test/test_cuda_nvml_based_avail.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    SUBPROCESS_REMINDER_MSG = (
+        "\n REMINDER: Tests defined in test_cuda_nvml_based_avail.py must be run in a process "
+        "where there CUDA Driver API has not been initialized. Before further debugging, ensure you are either using "
+        "run_test.py or have added --subprocess to run each test in a different subprocess.")
+
+    def setUp(self):
+        super().setUp()
+        torch.cuda.device_count.cache_clear()  # clear the lru_cache on this method before our test
+        if os.getenv('PYTORCH_NVML_BASED_CUDA_CHECK'):
+            del os.environ['PYTORCH_NVML_BASED_CUDA_CHECK']
```
Comment (ngimel) [inline test/test_cuda_nvml_based_avail.py:38]: should you restore os env in the teardown?

**Review by ngimel — COMMENTED** [review]
THis generally looks good, I left minor comments

File: docs/source/notes/cuda.rst (hunk 4 above)
Comment (speediedan) [reply docs/source/notes/cuda.rst:480]: Done!

File: test/test_cuda_nvml_based_avail.py (hunk 5 above)
Comment (speediedan) [reply test/test_cuda_nvml_based_avail.py]: I've now moved the os env config that was required for some of the test module imports behind a context guard so all os env modifications in this module (in the test case and outside of it) should be context guarded obviating the need for a teardown override.

Comment (speediedan) [reply test/test_cuda_nvml_based_avail.py]: Thanks again for the review and your help @ngimel! :rocket: :tada:

File: test/test_cuda_nvml_based_avail.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+# NOTE: Each of the tests in this module need to be run in a brand new process to ensure CUDA is uninitialized
+# prior to test initiation.
+with patch.dict(os.environ, {"PYTORCH_NVML_BASED_CUDA_CHECK": "1"}):
+    # Before executing the desired tests, we need to disable CUDA initialization and fork_handler additions that would
+    # otherwise be triggered by the `torch.testing._internal.common_utils` module import
+    from torch.testing._internal.common_utils import (parametrize, instantiate_parametrized_tests, run_tests, TestCase,
+                                                      IS_WINDOWS)
+    # NOTE: Because `remove_device_and_dtype_suffixes` initializes CUDA context (triggered via the import of
+    # `torch.testing._internal.common_device_type` which imports `torch.testing._internal.common_cuda`) we need
```
Comment (ngimel) [inline test/test_cuda_nvml_based_avail.py:18]: cc @mruberry, @pmeier this cuda context initialization seems unneeded (no action necessary on this PR @speediedan)

**Review by ngimel — APPROVED** [review]
Awesome, thanks @speediedan!
If tests are green, you can merge it yourself by leaving "@pytorchbot merge" comment, otherwise ping me.

File: test/test_cuda_nvml_based_avail.py (hunk 6 above)
Comment (mruberry) [reply test/test_cuda_nvml_based_avail.py:18]: That's true. This happens because `common_device_type.py` imports from `common_cuda.py`, which unconditionally initializes the CUDA context. We could probably refactor that import . Would you file an issue?

**Comment by speediedan** [comment]
> Awesome, thanks @speediedan! If tests are green, you can merge it yourself by leaving "@pytorchbot merge" comment, otherwise ping me.

Thanks again for all your help and work @ngimel! 

I noticed there was one failing test due to `test_cuda_nvml_based_avail.py` being included in ROCm testing (only 1 of the 6 tests requires functioning NVML for success). I could either:

1. add this test module to the ROCm block list OR
2. add another condition to the test that checks for actual NVML functioning and allows execution on ROCm:
    ```python
    if os.getenv('PYTORCH_NVML_BASED_CUDA_CHECK') == '1' and nvml_avail and torch.cuda._device_count_nvml() > 0:
    ```    
I went with option 1 in the most recent commit since I thought running these 6 tests in a ROCm context might be more wasteful than useful. All the checks are green as of that commit (except an unrelated [android emulator timeout](https://github.com/pytorch/pytorch/actions/runs/3231893793/jobs/5291898699#step:6:708)), but I'm happy to switch to option 2 if you want, just lmk how you prefer to proceed. 

Thanks again!

**Comment by ngimel** [comment]
I'm happy with option 1.

**Comment by speediedan** [comment]
> @pytorchbot merge -f "test timeout unrelated"

Awesome, thanks again @ngimel. Should we add [release notes: cuda](https://github.com/pytorch/pytorch/labels/release%20notes%3A%20cuda) and [topic: developer feature](https://github.com/pytorch/pytorch/labels/topic%3A%20developer%20feature) labels?

**Comment by ngimel** [comment]
Yep, adding labels would be great!
